### DISCUSSION TITLE
Proposal: Alternate syntax to !! for null parameter checks

### DISCUSSION THREAD

**Discussion by coolbluewater** [dotnet/csharplang#6100] [Language Ideas]
This is an alternate proposal to the proposed `!!` null parameter checks in C# 11 - see https://github.com/dotnet/runtime/pull/64720.

**Example 1: Regular method**
Instead of:
```C#
void M(object arg1!!, int arg3, object arg3!!)
{
    ...
}
```
Write:
```C#
void M(object arg1, int arg2, object arg3)
   : notnull(arg1, arg3)
{
    ...
}
```

**Example 2: Constructors**
Instead of: 
```C#
Class(object arg1!!, int arg2, object arg3!!) 
   : base(arg1, arg2)
{
    ...
}
```
Write:
```C#
Class(object arg1, int arg2, object arg3)
   : notnull(arg1, arg3)
   : base(arg1, arg2)
{
    ...
}
```

**Discussion**
1. The keyword is up for debate. Some other possibilities are `defines`, `defined`, `require`.
2. The syntax follows that of the `base` keyword, and so fits naturally into C#.
3. The placement of `notnull` indicates that it is executed before any of the code in the method.
4. For constructors, `notnull` must come before the `base` statement, if any. This indicates that `notnull` is executed before the base class constructor.
5. The `notnull` statement takes only one line, regardless of the number of arguments to be null checked. 
6. The use of a separate `notnull` statement allows a breakpoint to be set.
7. This syntax provides enough room for other such statements in the future.

@KathleenDollard @stephentoub

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by HaloFour** [comment]
Null parameter checking has been pulled from C# 11, largely due to the negative feedback and bikeshedding over the specific syntax:

https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-13.md#parameter-null-checking

**Comment by coolbluewater** [comment]
Why has this proposal been made undiscoverable, with the original issue locked, closed and turned into a discussion (within minutes!) where nobody is likely to find it?

Folks have stated that they've been looking for a good solution for many years - and now there's no interest in one?

@jcouv @RikkiGibson @333fred @AlFasGD

**Reply by Rekkonnect** [reply]
We've all participated in a conversation in a relevant discussion: #5735

And I specifically [proposed](https://github.com/dotnet/csharplang/discussions/5735#discussioncomment-2152434) my position like so.

Nothing more to be said on `!!`, as it got completely cancelled. The team has expressed their inconclusivity on the approach they're into taking in the future version that the feature will be introduced.

**Reply by HaloFour** [reply]
Discussions are the appropriate place to discuss these proposals, although alternative syntax suggestions for existing proposals will likely be directed to an existing discussion to keep the conversation in one place.  Issues are reserved for the language team members to track design and specification work.

For this discussion specifically, tons of variations of syntax have already been discussed.  My opinion is that there isn't enough interest in the feature in general to overcome the many differences of opinion regarding syntax.

**Reply by CyrusNajmabadi** [reply]
> Why has this proposal been made undiscoverable,

I don't know what you mean by 'made undiscoverable'.  

> with the original issue locked,

THe original issue was locked because issues are used to track proposals that are being implemented.  We decided not to implement the feature here, so the issue was locked as it serves no further purpose.

> and turned into a discussion (within minutes!) where nobody is likely to find it?

Discussions are where design happens for things that haven't yet been approved for being added to hte language.

**Reply by CyrusNajmabadi** [reply]
> Folks have stated that they've been looking for a good solution for many years - and now there's no interest in one?

There is certainly interest.  But whatever proposal would have to substantively move past the issues we found with all the syntactic forms we explored.  This proposal does not do that.  It effectively mirrors many of the syntax forms we looked at, with all the same concerns we felt were present with those as well.

**Reply by jnm2** [reply]
> Why has this proposal been made undiscoverable, with the original issue locked, closed and turned into a discussion (within minutes!) where nobody is likely to find it?

@coolbluewater Is there any clarification we can give in the issue template selector, the issue template itself, and the repo readme that would have let you know that we want you to create a discussion (not an issue) for the kind of content you initially posted?

**Reply by jnm2** [reply]
Never mind, I missed that this was transferred in from the runtime repo.

**Comment by coolbluewater** [comment]
@AlFasGD  @HaloFour - these answers are only more disquieting. And there's no comment on the actual proposal.

The [cancellation](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-13.md#parameter-null-checking) was due to the specific syntax proposed, not because the problem went away.

This proposal is a completely fresh take. That's what "alternative" means - and it takes only a couple of minutes of reading to know that.

**Reply by CyrusNajmabadi** [reply]
> This proposal is a completely fresh take

This syntax is in line with many of the syntaxes we considered.  The team effectively rejected going down this path as we found none of the syntax directions satisfying enough for all of us, and we felt that was unsatisfactory given the problem we wanted to solve.

**Reply by coolbluewater** [reply]
> This syntax is in line with many of the syntaxes we considered. 

Can you point to one syntax that it is truly in line with - particularly relative to the points mentioned in the proposal?

**Reply by CyrusNajmabadi** [reply]
Any of the syntaxes that involved placing hte check after the end of the parameters but before the method body.

**Reply by CyrusNajmabadi** [reply]
Our feeling wsa that dedicated syntax was not worth it.  There was still interest potentially in the future about allowing code/checking to happen before things like base calls.  But given that we'd likely want to support all statement forms, the feeling was more that we might go more like the VB route where  the base-call itself is just a statement.  By doing that, we can leverage all existing statements, then have the base-call, then call any other statements (now allowing access to 'this').

**Reply by coolbluewater** [reply]
> Any of the syntaxes that involved placing hte check after the end of the parameters but before the method body.

Can you provide one syntax to measure up to the points mentioned in this proposal?

**Reply by CyrusNajmabadi** [reply]
To the specific points:

> The keyword is up for debate. Some other possibilities are defines, defined, require.

This doesn't really matter here.  So it's not really a pro/con for the proposal.

> The syntax follows that of the base keyword, and so fits naturally into C#.

Not really important.  Matching an existing syntax isn't something that mattered to us wrt to the feature.

> The placement of notnull indicates that it is executed before any of the code in the method.

Nice to have.  Same as practically all syntaxes considered.

> For constructors, notnull must come before the base statement, if any. This indicates that notnull is executed before the base class constructor.

Nice to have.   Same as practically all syntaxes considered.

> The notnull statement takes only one line, regardless of the number of arguments to be null checked.

Same as effectly all syntaxes considered.  C# is not whitespace sensitive.  So all syntaxes could be written on one line.

> The use of a separate notnull statement allows a breakpoint to be set.

Not relevant.  We can make the debugger do whatever we want.  This would be true of any syntaxes we picked.

> This syntax provides enough room for other such statements in the future.

Not really a virtue.  That's a knock against the feature as it means now we have to design the space for all statements we want in this space.  This ties into what i was saying above that if we wanted this facility, it would make more sense to actually allow this to be in the method where you can already write any statement.  e.g. trying to align with `:this`/`:base` is more limiting than trying to align with statements.

**Reply by coolbluewater** [reply]
Can you provide one syntax for comparison?

**Reply by CyrusNajmabadi** [reply]
Like i said, this was in line with all of our discussions about using the space between the end of hte parameter list and the start of the method body.  THe points are about syntax in that space in general, not about any specific choice.  Providing a syntax to compare to is not relevant because we're not critiquing a particular syntax, but the space of syntaxes in that family.

**Reply by 333fred** [reply]
The most recent LDM meeting on the topic provided multiple. This is most similar to the `where` clause, which was extremely disliked by a majority of the LDM: https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-13.md#parameter-null-checking.

**Reply by coolbluewater** [reply]
@333fred - if you are saying you can judge this proposal based on your opinion of the `where` clause, I'd have to disagree. It is much better to see this in its own right - they are not even superficially similar. I've seen proposed C# features poorly evaluated in this way. The actual syntax and keystrokes are key (no pun intended.) But if you really do want to compare, why not do a full comparison? Be interested to see if the exercise proves revealing to us. But spending a couple of seconds hand waving a proposal away indicates little due diligence.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater  analysis was provided here: https://github.com/dotnet/csharplang/discussions/6100#discussioncomment-2707471.  However, please note the points in https://github.com/dotnet/csharplang/discussions/6100#discussioncomment-2707443 and https://github.com/dotnet/csharplang/discussions/6100#discussioncomment-2707480

**Reply by coolbluewater** [reply]
@CyrusNajmabadi  My comment was a response to @333fred's comment. I've already read what you wrote, it has little to do with that. So be a good chap and let others speak.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater Please keep things civil and abide by the [.net code of conduct](https://dotnetfoundation.org/about/code-of-conduct).   Specifically:

> Demonstrating empathy and kindness toward other people
> Being respectful of differing opinions, viewpoints, and experiences
> Giving and gracefully accepting constructive feedback

We allow anyone to speak and respond on topics raised and questions asked.  We do not suppress on topic views.  

--

WRT to this topic, i was addressing the proposal you were making using the similar criteria and concepts that were brought up when we discussed the syntactic options at length in our last meeting.  Thanks! :)

**Reply by coolbluewater** [reply]
@CyrusNajmabadi - that's my point exactly. I asked you to let others speak! Constantly interjecting with a link to the same content keeps others from speaking.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater *everyone* is welcome here to talk and respond on anything if it is on the topic of discussion.  We do not suppress on topic views here. Any further requests to have people not respond to your posts will not be permitted.  If you woudl like to discuss this specific moderation further (or the topic of moderation in general) please message me in Discord: discord.gg/csharp.  Handle: Metasyntactic.

**Reply by coolbluewater** [reply]
Interesting - all my points get marked off-topic. I was not asking you not to respond. I asked you to give others a chance to speak, which is completely in line with your own code of conduct. That means being sensitive to others. How much of this page's content is from you and how much from @333fred? I'd say the ratio is more than 100:1. So does the very rule you quoted not apply to you? It's just basic manners to give room to others. And now you're giving me a twitter (sorry, discord) handle to message you on? Is that normal policy?

**Reply by coolbluewater** [reply]
I am not comfortable speaking with you on the subject of moderation. Perhaps I can speak to your manager? What is their handle?

**Reply by CyrusNajmabadi** [reply]
>  And now you're giving me a twitter (sorry, discord) handle to message you on? Is that normal policy?

Yes.  Any external channels are ok to discuss moderation.  As before, personal discussion about moderation and the rules here itself will not happen on in this specific repo 'Discussion'.  You're welcome to take me up on this or not.  It will not impact future moderation, but it will likely help you out wrt to understanding the rules here.  Thanks :)

**Reply by 333fred** [reply]
> How much of this page's content is from you and how much from @333fred?

My notes are not from the perspective of any individual member of the LDT. Indeed, I regularly have to type up and argue for proposals that I disagree with in some fashion, but the broader LDT was convinced by. Cyrus is equally well-reflected in those notes as any other LDT member.

As to your specific proposal, I would say that it is extremely similar to the where clause version. It has a bit less typing, true, but not to the point of making a difference in my eyes. We disliked the need to repeat parameters in multiple locations, which your proposal requires as well. The where clause version is also significantly easier to generalize, as patterns are already a well-integrated part of the language, while this new initializer-like format does not seem to generalize well. So I'm afraid that I don't see this changing our minds about pulling the feature from C# 11.

**Reply by coolbluewater** [reply]
Well I've saved this file offline (and several previous versions.) A bit spooked here. Do people really have this much free time on their hands for all this nonsense? Good grief.

**Comment by CyrusNajmabadi** [comment]
@coolbluewater If you're interested, feel free to come discuss this with team members over at discord.gg/csharp.  Thanks! :)

**Reply by coolbluewater** [reply]
@KathleenDollard - could you please weigh in here? Stonewalling records are being set.

**Reply by Rekkonnect** [reply]
This behavior pretty much resembles a Karen asking for the manager to step in

Wanna keep talking about this thing? Join the Discord as already directed. Not interested in joining the Discord? Your opinions will have lost their value by your behavior of dodging. If anything, we're trying to find a commonly acceptable solution to a problem that affects all of us, C# users, so cooperation is key.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater please stop pinging random team members.  LDM members are engaging with you and sharing information on the topic already.  We've also invited you to directly converse with us here, or with our other communities.  

Discussions should stay on topic.  You are free to engage with the community here (or our other locations).  If others want to participate, they can do so.  You do not need to spam them to pull them in.

**Reply by coolbluewater** [reply]
> This behavior pretty much resembles a Karen asking for the manager to step in

I've asked Kathleen to step in because of her role as PM in the languages team. Are you saying she has no role here?
Using such language only lowers your own standard.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater As above:

> Discussions should stay on topic. You are free to engage with the community here (or our other locations). If others want to participate, they can do so. You do not need to spam them to pull them in.

Please do not ping random team members on topics.  It is not necessary.  The team already maintains and watches this repo.  People can participate in the discussions they feel are appropriate *themselves*.  

--

Put another way, imagine if everyone acted in this fashion, pinging the same people (be they PM, or some other LDM member) on every issue because they wanted them to weigh in.  It's spamming them and unnecessary.  If they want to engage, they can do so. 

As above, you are already engaging with LDM members (including those who were part of the decision on `!!`).  You can participate here or in our other forums and we will gladly talk to you about this.  Thank you.

**Reply by coolbluewater** [reply]
@CyrusNajmabadi  - please explain why you keep deleting my comments. I asked @KathleenDollard who is the PM on the languages team. Why is that not appropriate? She has (as far as I can tell) been quite involved in this area.

I don't speak in semaphore. You have not provided a syntax to compare against. Then you have deleted my comments in which I asked Kathleen who is directly involved in this area to participate in the discussion. How is that "random"? Please answer directly and don't try to put me in the wrong. I have said only what is completely proper.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater  As i've stated above (and marked the comments) they are spamming people.  Again, please consider what woudl happen if everyone did this.  This repo is already monitored by this team and we can pull in people as appropriate.  SO far you've pinged 7 different members on this topic.  You do not need to do this.  

> You have not provided a syntax to compare against. 

As i mentioned above: 

> Like i said, this was in line with all of our discussions about using the space between the end of hte parameter list and the start of the method body. THe points are about syntax in that space in general, not about any specific choice. Providing a syntax to compare to is not relevant because we're not critiquing a particular syntax, but the space of syntaxes in that family.

**Reply by CyrusNajmabadi** [reply]
To repeat, the position of the team was not: we like this category of syntax, but we just need someone to find the best syntax here.

The position was: we don't feel like this category is sufficiently worthwhile that any syntax in it would make this feature palatable.  

I explained the problems we had here, and even applied them to your points about your syntax.  It shares the same issues which makes it similarly unpalatable.

**Reply by coolbluewater** [reply]
I will say that what you've said above is an extreme reversal of the position you had taken just a few weeks ago.  My feeling at that time, quite honestly, was that the language design team had taken leave of not just taste, but also perspective, in being willing to go ahead despite all feedback to the contrary.

This proposal was posted not knowing that the original proposal had been cancelled. It was ready quite a while back but for the prospect of getting into an um, difficult discussion. That has been the case for several other unvoiced suggestions over time - the C# saloon has, from one outsiders' perspective, become a tad cliquish and averse to community interference, I mean help.

It is in any case a relief that wiser heads decided to drop the original proposal.

**Reply by CyrusNajmabadi** [reply]
> the C# saloon has, from one outsiders' perspective, become a tad cliquish and averse to community interference, I mean help.

We welcome outside perspective and help.  You are welcome at any time to post your proposals and discuss them with the team members and the community writ large. 

As with all things, those proposals will likely go places (or not) based on how well received they are and how much value the team thinks they will bring to the language and ecosystem.

> It is in any case a relief that wiser heads decided to drop the original proposal.

As above, this is the result precisely of community feedback.  It's hard to reconcile the belief that we are averse to community participation and feedback, while also noticing that we directly changed course on this because of the engage with the community here.

**Reply by CyrusNajmabadi** [reply]
> This proposal was posted not knowing that the original proposal had been cancelled

FYI, design notes are published here: https://github.com/dotnet/csharplang/tree/main/meetings/2022

And they are always posted as discussions (for example: https://github.com/dotnet/csharplang/discussions/6087).  

We make all our minutes known and participate before, during, and after these meetings/decisions.  Thanks!

**Comment by coolbluewater** [comment]
In closing, please see that nothing pleasant, receptive or friendly has been communicated at any point. It has all been combative. Crude language by @AlFasGD has not been even subtly apologized for. I don't need these things. You do. That is, if you are interested in a joyful, happy, intelligent and productive community. And they do exist.

**Reply by CyrusNajmabadi** [reply]
@coolbluewater Please reach out to @AlFasGD separately if you would like to engage with him on that topic.  This is the repo for csharplang, we are not responsible for, or in control of, individual community members.  

> In closing, please see that nothing pleasant, receptive or friendly has been communicated at any point.

Many people engaged you pleasantly and addressed your points and proposal.  I can see Halo and Fred did, as did alfasgd (who you pinged to participate).

> I don't need these things. You do. That is, if you are interested in a joyful, happy, intelligent and productive community.

@coolbluewater We have asked simply that you not spam people about your posts.  You are welcome to otherwise engage, and several people so far (in both the community and LDM) have engaged with you on your topic and how we feel about it.  None of that was combative.  It was simply an honest and accurate assessment of what has transpired so far, and how your proposal fits into the general thinking that has happened so far.  

--

As mentioned above, you might find you prefer engaging with one of our more interactive communities.  I recommend our discord channels for this as that direct interaction format there might be better suited to the brainstorming it feels like you are trying to engage in.

Cheers! :)

**Reply by Rekkonnect** [reply]
I have no association with the LDT for them to owe you any apologies for my "crude" language that wasn't even crude. The initial problem is your arrogant attitude, and despite all the efforts to control the discussion, you failed to respect the place. As a regular visitor on this repo and its conversations, such events are thankfully rare, further proving the importance of containing rage, seeing how other discussions have been flowing.

In closing, despite all the directions you were given, you still concluded that "nothing pleasant, receptive or friendly has been communicated at any point". Nobody even remotely offended you for your suggestion, and the main topic of this entire discussion has shifted into your inability to be cooperative, which I specifically noted in my first message in this discussion.

Authorities know better how to handle the situation, but the moment you brought my name in, I'm responding appropriately with all due respect to other members' presence. Let's put an end to unnecessary clogging of this discussion.

**Comment by coolbluewater** [comment]
From this perspective, that is not it. You seem unlikely to get it, sadly, so good luck. No replies needed, please.

**Comment by CyrusNajmabadi** [comment]
to @coolbluewater and any other community members here.  Please keep the discussion on topic to the idea in the original OP.  If you want to discuss other matters, including feelings between yourselves, please take it up through some external channel.  The csharplang repo is not the right location for this.  Thanks.

If you would like to discuss the meta commentary further, please come to discord.gg/csharp as it is better suited to discussions of moderation.  Thank you :)
